pkg/controller/secret_controller.go

```diff
@@ -0,0 +1,135 @@
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/juicedata/juicefs-csi-driver/pkg/config"
+	"github.com/juicedata/juicefs-csi-driver/pkg/juicefs"
+	"github.com/juicedata/juicefs-csi-driver/pkg/k8sclient"
+)
+
+type SecretController struct {
+	*k8sclient.K8sClient
+}
+
+func NewSecretController(client *k8sclient.K8sClient) *SecretController {
+	return &SecretController{client}
+}
+
+func (m *SecretController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	klog.V(6).Infof("Receive secret %s %s", request.Name, request.Namespace)
+	secrets, err := m.GetSecret(ctx, request.Name, request.Namespace)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		klog.Errorf("get secret %s error: %v", request.Name, err)
+		return reconcile.Result{}, err
+	}
+	if secrets == nil {
+		klog.V(6).Infof("secret %s has been deleted.", request.Name)
+		return reconcile.Result{}, nil
+	}
+	if _, found := secrets.Data["token"]; !found {
+		klog.V(6).Infof("token not found in secret %s", request.Name)
+		return reconcile.Result{}, nil
+	}
+	if _, found := secrets.Data["name"]; !found {
+		klog.V(6).Infof("name not found in secret %s", request.Name)
+		return reconcile.Result{}, nil
+	}
+	jfs := juicefs.NewJfsProvider(nil, nil)
+	secretsMap := make(map[string]string)
+	for k, v := range secrets.Data {
+		secretsMap[k] = string(v[:])
+	}
+	jfsSetting, err := jfs.Settings(ctx, "", secretsMap, nil, nil)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	output, err := jfs.AuthFs(ctx, secretsMap, jfsSetting, true)
+	klog.V(6).Infof("auth output: %s.", output)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	conf := jfsSetting.Name + ".conf"
+	confPath := filepath.Join(config.ClientConfPath, conf)
+	b, err := os.ReadFile(confPath)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	confs := string(b)
+	secretsMap["initConfig"] = confs
+	secrets.StringData = secretsMap
+	err = m.UpdateSecret(ctx, secrets)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (m *SecretController) SetupWithManager(mgr ctrl.Manager) error {
+	c, err := controller.New("mount", mgr, controller.Options{Reconciler: m})
```

Comment:
Controller name should be unique in the same manager.

Reply:
I've changed it to "secret"